Various fixes for running e2e tests

wallrj opened the pull request:
See https://github.com/jetstack/navigator/pull/45 for discussion

```release-note
NONE
```

wallrj commented:
In https://github.com/jetstack/navigator/pull/45#issuecomment-338957291 @munnerz wrote:
> Can we switch this back? These test two different things - the api may be available whilst the controller is in a crash loop.

I don't mind putting it back but my thinking is that we shouldn't be poking around for the state of controller in e2e tests...unless the tests fail.
If the Navigator API is ready we should be able to create a Navigator database and then expect the database to start up in some <short> time.

And in any case, the controller is running in the tests right now, but failing because RBAC policy prevents its leader election routine to run.


```
ERROR: logging before flag.Parse: I1030 17:40:31.035544       1 round_trippers.go:417] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: navigator-controller/v1.8.2 (linux/amd64) kubernetes/$Format/leader-election" -H "Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6Im5hdi1lMmUtbmF2aWdhdG9yLWNvbnRyb2xsZXItdG9rZW4tenRsNDUiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoibmF2LWUyZS1uYXZpZ2F0b3ItY29udHJvbGxlciIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6ImI5MzM3ZmI3LWJkOTYtMTFlNy04ZWQ4LTUyNTQwMDg2NzBlYyIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpkZWZhdWx0Om5hdi1lMmUtbmF2aWdhdG9yLWNvbnRyb2xsZXIifQ.rq5OBsceqNzikBLbjIViy-yk1A22nn-dfRuRF_MWZiFjbRwBmMe4ZACk2O06mSPb-GaDvHS6ryeAaEwXNQZB_cyIKKgxPUabcRprgTh0-Ghl6K2w4d77s2gdERb-yBgRjffaa1QGAj_n8M0MbAGOVfHPvs4x8M83QnjIDfwmkIDw0u_-GboOWS1qKyb42sU3tFu7ByoMPqvlV7VX5gXmdJWcSKBoyv7GkoNFP_1bNp_2NieCC_XgmcmGAHOMxUgawFD4idtrMr3I3ReCIoC_p_mKJyqMaQKyNbuINRQ72lqgna3ZgCg7Nlo2h8eqoFvxOvKFWnsPA5HzAeRE7jxJRA" https://10.0.0.1:443/api/v1/namespaces/kube-system/endpoints/navigator-controller
ERROR: logging before flag.Parse: I1030 17:40:31.158865       1 round_trippers.go:436] GET https://10.0.0.1:443/api/v1/namespaces/kube-system/endpoints/navigator-controller 403 Forbidden in 123 milliseconds
ERROR: logging before flag.Parse: I1030 17:40:31.158889       1 round_trippers.go:442] Response Headers:
ERROR: logging before flag.Parse: I1030 17:40:31.158894       1 round_trippers.go:445]     Date: Mon, 30 Oct 2017 17:40:31 GMT
ERROR: logging before flag.Parse: I1030 17:40:31.158898       1 round_trippers.go:445]     Content-Type: text/plain
ERROR: logging before flag.Parse: I1030 17:40:31.158901       1 round_trippers.go:445]     X-Content-Type-Options: nosniff
ERROR: logging before flag.Parse: I1030 17:40:31.158905       1 round_trippers.go:445]     Content-Length: 118
ERROR: logging before flag.Parse: I1030 17:40:31.158928       1 request.go:836] Response Body: User "system:serviceaccount:default:nav-e2e-navigator-controller" cannot get endpoints in the namespace "kube-system".
ERROR: logging before flag.Parse: E1030 17:40:31.158961       1 leaderelection.go:224] error retrieving resource lock kube-system/navigator-controller: User "system:serviceaccount:default:nav-e2e-navigator-controller" cannot get endpoints in the namespace "kube-system". (get endpoints navigator-controller)
ERROR: logging before flag.Parse: I1030 17:40:31.158970       1 leaderelection.go:180] failed to acquire lease kube-system/navigator-controller
```

We should fix that.

munnerz commented:
@wallrj agreed that it'd be nice to do a full test like that, but right now we don't have the testing infra to stand up full DB clusters due to resource limits on Travis. I'm working to fix this via our new test-infra, but for the time being, it's important we know whether the controller will work with the default configuration, so I'd like to keep it in.

Yep we should fix that (#68), and also should update the tests so they fail as a result of the leader election failing (through a health check perhaps). But this second part is a bit lower priority, as failing leader election is unlikely to be the cause of test failures (once we fix #68)

wallrj commented on .travis.yml:
```diff
@@ -18,9 +18,6 @@ jobs:
       - sudo -E CHANGE_MINIKUBE_NONE_USER=true minikube start --vm-driver=none --kubernetes-version="$KUBERNETES_VERSION" --extra-config=apiserver.Authorization.Mode=RBAC
-      - while true; do if kubectl get nodes; then break; fi; echo "Waiting 5s for kubernetes to be ready..."; sleep 5; done
-      # Fix problems with kube-dns + rbac
-      - kubectl create serviceaccount --namespace kube-system kube-dns
```
These steps are already performed in the prepare-e2e.sh script.

wallrj commented on hack/e2e.sh:
```diff
@@ -19,18 +19,12 @@ source "${SCRIPT_DIR}/libe2e.sh"
-retry TIMEOUT=600 kubectl get nodes
-
-echo "Waiting for tiller to be ready..."
-retry TIMEOUT=60 helm version
```
And these steps are also performed in prepare-e2e.sh, so I removed them.

wallrj commented on hack/e2e.sh:
```diff
@@ -44,58 +38,42 @@ function navigator_ready() {
-    exit 1
+    FAILURE_COUNT=$(($FAILURE_COUNT+1))
+    echo "TEST FAILURE: $1"
 }
```
This is so that *all* the tests are performed rather failing on the first test.

wallrj commented on hack/e2e.sh:
```diff
@@ -106,6 +84,11 @@ function test_elasticsearchcluster_create() {
+         --namespace "${USER_NAMESPACE}" \
+         service es-demo; then
+        fail_test "Navigator controller failed to create elasticsearchcluster service"
+    fi
```
Added a check to see that a service is created....a clear indication that the navigator controller is reacting to the new escluster resource.

wallrj commented on hack/prepare-e2e.sh:
```diff
@@ -1,25 +1,11 @@
-CERT_DIR="$CONFIG_DIR/certs"
-mkdir -p $CERT_DIR
-TEST_DIR="$CONFIG_DIR/tmp"
-mkdir -p $TEST_DIR
```
None of this stuff is needed in this script.

wallrj commented on hack/prepare-e2e.sh:
```diff
@@ -28,7 +14,13 @@ apiVersion: v1
+  kind: ServiceAccount
+  metadata:
+    name: kube-dns
+    namespace: kube-system
```
Added this here rather than creating the service in the travis.yml file.

munnerz replied:
My reasoning behind the separation is so that we are not opinionated in our e2e test scripts about using minikube. It'd be ideal if the e2e tests could be run against *any* valid/functioning kubernetes cluster (this is more relevant when it comes to us running our tests on GCE/GKE).

Happy to accept this for now, but I think there'll be more refactoring of this stuff soon 😄

wallrj replied:
Ah, got it. Well thanks for merging this anyway.

wallrj commented:
@munnerz Ok. I've added back the check for apiserver and controller replicas and added another check for `kubectl get esc` which was also failing when I run the tests locally.

I know you're planning to create e2e tests with Ginkgo / prow / gce, but I implore you to merge this change so that I can continue to perform some quick local e2e tests on my laptop.

wallrj commented:
e2e test failed because of failed `kubectl get esc` which I suspect was caused by  kube-dns failing:
 * https://travis-ci.org/jetstack/navigator/jobs/295687871#L853

```
+kubectl get pods --all-namespaces

NAMESPACE     NAME                                                                 READY     STATUS             RESTARTS   AGE

default       nav-e2e-navigator-apiserver-4058778657-jdmx9                         2/2       Running            0          1m

default       nav-e2e-navigator-controller-2991077948-7l4s5                        1/1       Running            0          1m

kube-system   kube-addon-manager-travis-job-345a1919-cc3d-44d6-8177-e50007723a0a   1/1       Running            0          8m

kube-system   kube-dns-1326421443-q1p8c                                            2/3       CrashLoopBackOff   7          8m

kube-system   kubernetes-dashboard-qb2lw                                           0/1       CrashLoopBackOff   6          8m

kube-system   tiller-deploy-3341511835-k8zxt                                       1/1       Running            0          1m
```

/retest

munnerz commented:
@wallrj /retest does not work for travis jobs :(

I've hit restart. Are you able to restart jobs on travis?

wallrj commented:
Ah yes, I see it here: https://travis-ci.org/jetstack/navigator/pull_requests 

Thanks @munnerz

munnerz commented:
/lgtm
/approve
